localstack/testing/scenario/provisioning.py

```diff
@@ -16,85 +23,168 @@
+        """
+        Execute all previously added custom provisioning steps and deploy added CDK stacks via CloudFormation.
+
+        Already deployed stacks will be updated instead.
+
+        """
+        is_update = False
+
         if all(
             self._is_stack_deployed(stack_name, stack)
             for stack_name, stack in self.cloudformation_stacks.items()
         ):
-            # TODO it's currently all or nothing -> deploying one new stack will most likely fail
-            LOG.info("All stacks are already deployed. Skipping the provisioning.")
-            self.skipped_provisioning = True
-            return
+            LOG.debug("All stacks are already deployed. Skipping the provisioning.")
+            # TODO: in localstack we might want to do a delete/create
+            #  but generally this won't be a common use case when developing against LocalStack
+            is_update = True
 
-        self.run_manual_setup_tasks()
-        self.bootstrap_cdk()
+        self._run_manual_setup_tasks()
+        self._bootstrap_cdk()
         for stack_name, stack in self.cloudformation_stacks.items():
-            self.aws_client.cloudformation.create_stack(
-                StackName=stack_name,
-                TemplateBody=stack["Template"],
-                Capabilities=[
-                    Capability.CAPABILITY_AUTO_EXPAND,
-                    Capability.CAPABILITY_IAM,
-                    Capability.CAPABILITY_NAMED_IAM,
-                ],
-            )
-            self.aws_client.cloudformation.get_waiter("stack_create_complete").wait(
-                StackName=stack_name,
-                WaiterConfig=WAITER_CONFIG_AWS if is_aws_cloud() else WAITER_CONFIG_LS,
-            )
-            describe_stack = self.aws_client.cloudformation.describe_stacks(StackName=stack_name)
-            outputs = describe_stack["Stacks"][0].get("Outputs", {})
-            stack["Outputs"] = {o["OutputKey"]: o["OutputValue"] for o in outputs}
+            change_set_name = f"test-cs-{short_uid()}"
+            if len(stack["Template"]) > 51_200 or True:
+                # upload to s3 instead
+                template_bucket_name = self._template_bucket_name()
+                self._create_if_not_exists(template_bucket_name)
+                key = f"{stack_name}.yaml"
+                self.aws_client.s3.put_object(
+                    Bucket=template_bucket_name, Key=key, Body=stack["Template"]
+                )
+                url = self.aws_client.s3.generate_presigned_url(
+                    ClientMethod="get_object",
+                    Params={"Bucket": template_bucket_name, "Key": key},
+                    ExpiresIn=1600,
+                )
+                # domain = "amazonaws.com" if is_aws_cloud() else f"localhost:{config.EDGE_PORT}"
+                # url = f"https://{template_bucket_name}.s3.{domain}/{key}"
+                # f"https://s3.us-east-1.amazonaws.com/{template_bucket_name}/{stack_name}.yaml"
```

Comment:
what is the issue with S3 in that case? shouldn't the pre-signed URL use the endpoint configured in the client? what is the problematic behaviour?

Reply:
nothing actually, was an older issue